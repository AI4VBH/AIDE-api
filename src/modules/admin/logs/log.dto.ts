export class LogsDTO {
  json: LogDTO;

  public static from(dto: Partial<LogsDTO>) {
    const it = new LogsDTO();
    it.json = dto.json;
    return it;
  }

  // TODO: Implement fromEntity method
  // public static fromEntity(entity: Logs) {
  //   return this.from({
  //     task_id: entity.id,
  //   });
  // }

  // TODO: Implement toEntity method
  // public toEntity(logs: Logs = null) {
  //   const it = new Logs();
  //   it.task_id = this.task_id;
  //   return it;
  // }
}

export class LogDTO {
  execution_id: string;
  level: string;
  line_no: number;
  logger: string;
  model_name: string;
  model_version: string;
  module: string;
  msg: string;
  thread: string;
  type: string;
  written_at: string;
  written_ts: number;

  public static from(dto: Partial<LogDTO>) {
    const it = new LogDTO();
    it.execution_id = dto.execution_id;
    it.level = dto.level;
    it.line_no = dto.line_no;
    it.logger = dto.logger;
    it.model_name = dto.model_name;
    it.model_version = dto.model_version;
    it.module = dto.module;
    it.msg = dto.msg;
    it.thread = dto.thread;
    it.type = dto.type;
    it.written_at = dto.written_at;
    it.written_ts = dto.written_ts;
    return it;
  }

  // TODO: Implement fromEntity method
  // public static fromEntity(entity: Log) {
  //   return this.from({
  //     task_id: entity.id,
  //   });
  // }

  // TODO: Implement toEntity method
  // public toEntity(log: Log = null) {
  //   const it = new Log();
  //   it.task_id = this.task_id;
  //   return it;
  // }
}
