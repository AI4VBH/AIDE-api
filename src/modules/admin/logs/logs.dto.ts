export class LogsDto {
  message: string;
  correlationId: string;
  taskStatus: string;
  workflowInstanceId: string;
  task: unknown;
  timestamp: Date;
  level: string;
}
