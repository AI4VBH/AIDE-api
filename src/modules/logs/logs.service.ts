import { Injectable } from '@nestjs/common';
import { Log } from './log.interface';

@Injectable()
export class LogsService {
  getLog(): Log {
    return {} as Log;
  }
}
