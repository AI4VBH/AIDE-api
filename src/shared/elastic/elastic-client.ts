/*
 * Copyright 2022 Crown Copyright
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { Inject, Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { Client, ClientOptions } from '@opensearch-project/opensearch';
import { parseBoolean } from 'shared/util/parseBoolean';

@Injectable()
export class ElasticClient extends Client {
  searchIndex: string;
  public static createNode(config: ConfigService): string {
    const ssl = parseBoolean(config.get('ELASTIC_USE_SSL'));
    const host = config.get<string>('ELASTIC_HOST');
    const port = config.get('ELASTIC_PORT');
    const prefix = ssl ? 'https' : 'http';

    return `${prefix}://${host}:${port}`;
  }

  constructor(@Inject(ConfigService) config: ConfigService) {
    super({
      node: ElasticClient.createNode(config),
      auth: {
        username: config.get<string>('ELASTIC_USERNAME'),
        password: config.get<string>('ELASTIC_PASSWORD'),
      },
      ssl: {
        rejectUnauthorized: false,
        requestCert: false,
      },
    } as ClientOptions);
    this.searchIndex = config.get<string>('ELASTIC_INDEX');
  }

  public searchExecution(id: string) {
    return { query: { match: { 'task.ExecutionId': id } } };
  }

  public async getLogs(id: string) {
    try {
      return await this.search({
        index: this.searchIndex,
        body: this.searchExecution(id),
      });
    } catch (error) {
      throw new ElasticClientException(error);
    }
  }
}

export class ElasticClientException extends Error {
  constructor(message: string) {
    super(message);
  }
}
