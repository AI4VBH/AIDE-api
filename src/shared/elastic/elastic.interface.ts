export interface Executions {
  hits: {
    total: number;
    hits: Array<{
      _source: any;
    }>;
  };
}

export interface ExecutionsSearchBody {
  id: number;
  title: string;
  content: string;
  authorId: number;
}

export interface ExecutionsSearchResult {
  hits: {
    total: number;
    hits: Array<{
      _source: any;
    }>;
  };
}
