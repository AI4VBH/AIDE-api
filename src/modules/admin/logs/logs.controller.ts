/*
 * Copyright 2022 Guy’s and St Thomas’ NHS Foundation Trust
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import {
  BadRequestException,
  Controller,
  Get,
  Param,
  UseFilters,
} from '@nestjs/common';
import ExternalServerExceptionFilter from 'shared/http/external-server-exception.filter';
import { LogsService } from './logs.service';

@Controller('logs')
@UseFilters(ExternalServerExceptionFilter)
export class LogsController {
  constructor(private readonly appService: LogsService) {}

  @Get(':id')
  getLogs(@Param('id') id: string) {
    if (!id) {
      throw new BadRequestException('id param cannot be empty');
    }
    return this.appService.getLogByTask(id);
  }
}
