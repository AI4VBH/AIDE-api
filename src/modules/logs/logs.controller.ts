import { Controller, Get } from '@nestjs/common';
import { Log } from './log.interface';
import { LogsService } from './logs.service';

@Controller('logs')
export class LogsController {
  constructor(private readonly appService: LogsService) {}

  @Get()
  getHello(): Log {
    return this.appService.getLog();
  }
}
