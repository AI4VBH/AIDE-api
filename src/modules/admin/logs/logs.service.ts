/*
 * Copyright 2022 Crown Copyright
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { Inject, Injectable } from '@nestjs/common';
import { ElasticClient } from 'shared/elastic/elastic-client';
import { LogsDto } from './logs.dto';
import { IElasticLogObject } from './models/logs.interfaces';

@Injectable()
export class LogsService {
  @Inject(ElasticClient)
  private readonly elasticClient: ElasticClient;

  async getLogByTask(id: string): Promise<LogsDto[]> {
    const response = await this.elasticClient.getLogs(id);

    const body = response.body as IElasticLogObject;

    return body.hits.hits.map(({ _source }) => {
      return {
        correlationId: _source['CorrelationId'],
        workflowInstanceId: _source['workflowInstanceId'],
        task: _source['task'],
        message: _source['Message'],
        taskStatus: _source['taskStatus'],
        level: _source['Level'],
        timestamp: _source['@timestamp'],
      } as LogsDto;
    });
  }
}
