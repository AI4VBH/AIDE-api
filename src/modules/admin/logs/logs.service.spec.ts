/*
 * Copyright 2022 Crown Copyright
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { DeepMocked, createMock } from '@golevelup/ts-jest';
import { Test, TestingModule } from '@nestjs/testing';
import { ElasticClient } from 'shared/elastic/elastic-client';
import { LogsService } from './logs.service';

describe('LogsService', () => {
  let service: LogsService;
  let elasticClient: DeepMocked<ElasticClient>;

  beforeEach(async () => {
    elasticClient = createMock<ElasticClient>();
    service = createMock<LogsService>();

    const module: TestingModule = await Test.createTestingModule({
      providers: [
        LogsService,
        { provide: ElasticClient, useValue: elasticClient },
      ],
    }).compile();

    service = module.get<LogsService>(LogsService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('get logs by task', () => {
    it('returns expected result', async () => {
      const response = { statusCode: 200, body: elasticResponseObj } as any;
      elasticClient.getLogs.mockResolvedValue(response);

      const result = await service.getLogByTask(
        '59b14f66-4a5f-41a1-8e9f-acd706413857',
      );

      expect(elasticClient.getLogs).toHaveBeenCalled();
      expect(result).toMatchSnapshot();
    });
  });

  const elasticResponseObj = {
    took: 31,
    timed_out: false,
    _shards: {
      total: 42,
      successful: 42,
      skipped: 0,
      failed: 0,
    },
    hits: {
      total: {
        value: 10000,
        relation: 'gte',
      },
      max_score: 45.33331,
      hits: [
        {
          _index: 'logstash-2022.11.22',
          _type: '_doc',
          _id: 'Hl2snoQBNksrDXBpRxQ8',
          _score: 45.33331,
          _source: {
            path: '%{path.Value}',
            argoNamespace: 'argo',
            MessageId: 'd6f39216-08cf-4cc1-b11a-fd60774783ab',
            Level: 'DEBUG',
            ServiceName: 'Monai.Deploy.WorkflowManager.TaskManager',
            EventName: 'CreatingArgoClient',
            messageType: 'TaskDispatchEvent',
            CorrelationId: 'a65f65e6-0c68-46c5-aa97-79084a662a21',
            EventId: 1006,
            RecievedTime: '2022-11-22T09:29:50.3068175Z',
            LoggerName:
              'Monai.Deploy.WorkflowManager.TaskManager.Argo.ArgoProvider',
            executionId: 'a6346d13-f8ab-43ec-843c-51f0d560c812',
            ServiceVersion: '0.0.0-development',
            '@timestamp': '2022-11-22T09:30:01.285Z',
            workflowInstanceId: 'ec1c0d79-1209-4067-9ef7-d384fb5d21f6',
            Message:
              'Creating Argo client with base URL: http://argo-argo-workflows-server.argo:2746.',
            type: 'nlog',
            ApplicationId: '16988a78-87b5-4168-a5c3-2cfc2bab8e54',
            taskId: 'mean-pixel-calc',
            MachineName: 'task-manager-6f78f55875-6p6fq',
            baseUrl: 'http://argo-argo-workflows-server.argo:2746',
            '@version': '1',
            headers: {
              connection: 'Keep-Alive',
              http_user_agent: null,
              content_length: '928',
              http_accept: null,
              request_path: '/',
              http_version: 'HTTP/1.1',
              http_host: 'logstash.shared:5011',
              request_method: 'POST',
            },
          },
        },
        {
          _index: 'logstash-2022.11.22',
          _type: '_doc',
          _id: 'Il2snoQBNksrDXBpSBQT',
          _score: 45.33331,
          _source: {
            path: '%{path.Value}',
            argoNamespace: 'argo',
            MessageId: 'd6f39216-08cf-4cc1-b11a-fd60774783ab',
            Level: 'DEBUG',
            ServiceName: 'Monai.Deploy.WorkflowManager.TaskManager',
            EventName: 'GeneratingArtifactSecret',
            messageType: 'TaskDispatchEvent',
            CorrelationId: 'a65f65e6-0c68-46c5-aa97-79084a662a21',
            EventId: 1000,
            RecievedTime: '2022-11-22T09:29:50.3068175Z',
            LoggerName:
              'Monai.Deploy.WorkflowManager.TaskManager.Argo.ArgoPlugin',
            executionId: 'a6346d13-f8ab-43ec-843c-51f0d560c812',
            ServiceVersion: '0.0.0-development',
            '@timestamp': '2022-11-22T09:30:01.492Z',
            name: 'input-dicom',
            Message:
              'Generating Kubernetes secrets for accessing artifacts: input-dicom.',
            type: 'nlog',
            ApplicationId: '16988a78-87b5-4168-a5c3-2cfc2bab8e54',
            workflowInstanceId: 'ec1c0d79-1209-4067-9ef7-d384fb5d21f6',
            taskId: 'mean-pixel-calc',
            MachineName: 'task-manager-6f78f55875-6p6fq',
            '@version': '1',
            headers: {
              connection: 'Keep-Alive',
              http_user_agent: null,
              content_length: '884',
              http_accept: null,
              request_path: '/',
              http_version: 'HTTP/1.1',
              http_host: 'logstash.shared:5011',
              request_method: 'POST',
            },
          },
        },
      ],
    },
  };
});
