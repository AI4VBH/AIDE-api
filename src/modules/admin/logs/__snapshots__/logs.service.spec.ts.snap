// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogsService get logs by task returns expected result 1`] = `
[
  {
    "correlationId": "a65f65e6-0c68-46c5-aa97-79084a662a21",
    "level": "DEBUG",
    "message": "Creating Argo client with base URL: http://argo-argo-workflows-server.argo:2746.",
    "taskId": "mean-pixel-calc",
    "taskStatus": undefined,
    "timestamp": "2022-11-22T09:30:01.285Z",
    "workflowInstanceId": "ec1c0d79-1209-4067-9ef7-d384fb5d21f6",
  },
  {
    "correlationId": "a65f65e6-0c68-46c5-aa97-79084a662a21",
    "level": "DEBUG",
    "message": "Generating Kubernetes secrets for accessing artifacts: input-dicom.",
    "taskId": "mean-pixel-calc",
    "taskStatus": undefined,
    "timestamp": "2022-11-22T09:30:01.492Z",
    "workflowInstanceId": "ec1c0d79-1209-4067-9ef7-d384fb5d21f6",
  },
]
`;
